attribution/experiment_logger.py

```diff
@@ -0,0 +1,258 @@
+        self.df_input_token_attribution.loc[len(self.df_input_token_attribution)] = {
+            "exp_id": self.experiment_id,
+            "attribution_strategy": attribution_strategy,
+            "input_token_pos": token_pos,
+            "input_token": token,
+            "attr_score": attr_score,
+        }
+
+    def log_token_attribution_matrix(
+        self,
+        attribution_strategy: str,
+        input_token_pos: int,
+        output_token_pos: int,
+        output_token: str,
+        attr_score: float,
+    ):
+        self.df_token_attribution_matrix.loc[len(self.df_token_attribution_matrix)] = {
+            "exp_id": self.experiment_id,
+            "attribution_strategy": attribution_strategy,
+            "input_token_pos": input_token_pos,
+            "output_token_pos": output_token_pos,
+            "output_token": output_token,
+            "attr_score": attr_score,
+        }
+
+    def log_perturbation(
+        self,
+        perturbation_pos: int,
+        perturbation_token: str,
+        perturbation_strategy: str,
+        original_input: str,
+        original_output: str,
+        perturbed_input: str,
+        perturbed_output: str,
+    ):
+        self.df_perturbations.loc[len(self.df_perturbations)] = {
+            "exp_id": self.experiment_id,
+            "perturbation_pos": perturbation_pos,
+            "perturbation_token": perturbation_token,
+            "perturbation_strategy": perturbation_strategy,
+            "original_input": original_input,
+            "original_output": original_output,
+            "perturbed_input": perturbed_input,
+            "perturbed_output": perturbed_output,
+        }
+
+    def clean_tokens(self, tokens):
+        return [token.replace("Ġ", "") for token in tokens]
+
+    def print_sentence_attribution(self):
+        sentences = []
+
+        for (exp_id, attr_strat), exp_data in self.df_input_token_attribution.groupby(
+            ["exp_id", "attribution_strategy"]
+        ):
+            tokens = self.clean_tokens(exp_data["input_token"].tolist())
+            attr_scores = exp_data["attr_score"].tolist()
+
+            token_attrs = [
+                f"{token}\n{score:.2f}" for token, score in zip(tokens, attr_scores)
+            ]
+
+            perturbation_strategy = self.df_experiments.loc[
+                self.df_experiments["exp_id"] == exp_id, "perturbation_strategy"
+            ].values[0]
+
+            sentence_data = {
+                "exp_id": exp_id,
+                "attribution_strategy": attr_strat,
+                "perturbation_strategy": perturbation_strategy,
+            }
+            sentence_data.update(
+                {f"token_{i+1}": token_attr for i, token_attr in enumerate(token_attrs)}
+            )
+
+            sentences.append(sentence_data)
+
+        df_sentences = pd.DataFrame(sentences)
+        self.pretty_print(df_sentences)
+
+    def print_attribution_matrix(self, exp_id: int, attribution_strategy: str):
```

Comment:
I think this would be easier if there was an option to take the row index, instead of having to specify the experiment number and attribution strategy. Or, if just the experiment number is provided, print all.

Reply:
One option is to ask the `exp_id` and make the `attribution_strategy` optional, and default to print another one. I'll make that.

Another option would be to ask the `row_id` instead, but that would need to be a different method. Let me know if you're interested in having that as well!

Reply:
First option sounds good, thanks!